File: pkg/src/polymat.c
/* $ID: polymat, last updated 2022-07-04, F.Osorio */

#include "fastmatrix.h"

void
matrix_polynomial(double *a, int *lda, int *n, double *coef, int *ncoef, double *b, int *ldb, int *info)
{ /* evaluates a real general matrix polynomial using a Horner's scheme */
  FM_matrix_pol(a, *lda, *n, coef, *ncoef, b, *ldb, info);
}

void
FM_matrix_pol(double *a, int lda, int n, double *coef, int ncoef, double *b, int ldb, int *info)
{ /* evaluates a real general matrix polynomial */

  /* test the input parameters */
  *info = 0;
  if (n < 0) {
    *info = 3;
  } else if (lda < MAX(1, n)) {
    *info = 2;
  } else if (ncoef < 0) {
    *info = 5;
  } else if (ldb < MAX(1, n)) {
    *info = 7;
  }
  if (*info != 0) return;

  /* quick return if possible */
  if (n == 0)
    return;

  if (ncoef == 0) {
    for (int j = 0; j < n; j++)
      b[j * (n + 1)] = coef[0];
    return;
  }

  /* Horner's scheme */
  for (int j = 0; j < n; j++)
    b[j * (n + 1)] = coef[ncoef];
  while (ncoef-- > 0) {
    FM_mult_mat(b, a, lda, n, n, b, ldb, n, n);
    for (int j = 0; j < n; j++)
      b[j * (n + 1)] += coef[ncoef];
  }
}

File: pkg/R/polymat.R
## ID: polymat.R, last updated 2022-08-08, F.Osorio

matrix.polynomial <- function(a, coef = rep(1, power + 1), power = length(coef))
{ ## evaluates a real general matrix polynomial using a Horner's scheme

  # validating arguments
  if (is.data.frame(a))
    a <- as.matrix(a)
  if (!is.matrix(a))
    stop("supply a matrix-like 'a'")
  if (!is.numeric(a))
    stop("argument 'a' is not a numeric matrix")
  da <- dim(a)
  n <- da[1]
  if (n != da[2])
    stop("argument 'a' is not a square matrix")
  storage.mode(a) <- "double"

  if (!is.vector(coef))
    coef <- as.vector(coef)
  if (!is.numeric(coef))
    stop("argument 'coef' is not a numeric vector")
  ncoef <- length(coef)

  if (all(coef == 1.0) && (power < 0))
    stop("argument 'power' must be positive")

  mat <- matrix(0, nrow = n, ncol = n)
  storage.mode(mat) <- "double"
  z <- .C("matrix_polynomial",
          a = a,
          lda = as.integer(n),
          n = as.integer(n),
          coef = as.double(coef),
          ncoef = as.integer(ncoef),
          mat = mat,
          ldmat = as.integer(n),
          info = as.integer(0))[c("mat", "info")]
  if (z$info)
    stop(paste("matrix_polynomial gave error code", z$info))
  z <- z$mat
  z
}
